neps/nep-0413.md

```diff
@@ -0,0 +1,174 @@
+---
+NEP: 413
+Title: Near Wallet API - support for signMessage method
+Author: Philip Obosi <philip@near.org>, Guillermo Gallardo <guillermo@near.org>
+# DiscussionsTo:
+Status: Approved
+Type: Standards Track
+Category: Wallet
+Created: 25-Oct-2022
+---
+
+## Summary
+
+A standardized Wallet API method, namely `signMessage`, that allows users to sign a message for a specific recipient using their NEAR account.
+
+## Motivation
+NEAR users want to create messages destined to a specific recipient using their accounts. This has multiple applications, one of them being authentication in third-party services.
+
+Currently, there is no standardized way for wallets to sign a message destined to a specific recipient.
+
+## Rationale and Alternatives
+Users want to sign messages for a specific recipient without incurring in GAS fees, nor compromising their account's security. This means that the message being signed:
+
+1) Must be signed off-chain, with no transactions being involved.
+2) Must include the recipient's name and a nonce.
+3) Cannot represent a valid transaction.
+3) Must be signed using a Full Access Key.
+4) Should be simple to produce/verify, and transmitted securely.
+
+### Why Off-Chain?
+So the user would not incur in GAS fees, nor the signed message gets broadcasted into a public network.
+
+### Why The Message MUST NOT be a Transaction? How To Ensure This?
+An attacker could make the user inadvertently sign a valid transaction which, once signed, could be submitted into the network to execute it.
+
+#### How to Ensure the Message is not a Transaction
+In NEAR, transactions are encoded in Borsh before being signed. The first attribute of a transaction is a `signerId: string`, which is encoded as: (1) 4 bytes representing the string's length, (2) N bytes representing the string itself.
+
+By prepending the prefix tag $2^{31} + 413$ we can both ensure that (1) the whole message is an invalid transaction (since the string would be too long to parse), (2) this NEP is ready for a potential future protocol update, in which non-consensus messages are tagged using $2^{31}$ + NEP-number.
```

Comment:
```suggestion
By prepending the prefix tag $2^{31} + 413$ we can both ensure that (1) the whole message is an invalid transaction (since the string would be too long to be a valid signer account id), (2) this NEP is ready for a potential future protocol update, in which non-consensus messages are tagged using $2^{31}$ + NEP-number.
```